core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -438,12 +439,27 @@ class BlockManagerMasterEndpoint(
     if (blockLocations.containsKey(blockId)) blockLocations.get(blockId).toSeq else Seq.empty
   }
 
-  private def getLocationsAndStatus(blockId: BlockId): Option[BlockLocationsAndStatus] = {
+  private def getLocationsAndStatus(
+      blockId: BlockId, requesterHost: String): Option[BlockLocationsAndStatus] = {
     val locations = Option(blockLocations.get(blockId)).map(_.toSeq).getOrElse(Seq.empty)
     val status = locations.headOption.flatMap { bmId => blockManagerInfo(bmId).getStatus(blockId) }
 
     if (locations.nonEmpty && status.isDefined) {
-      Some(BlockLocationsAndStatus(locations, status.get))
+      val bmIdToLocalDirs = if (status.get.storageLevel.useDisk) {
+        locations
+          .find(_.host == requesterHost)
+          .map { sameHostBlockId =>
+            val bmInfo = blockManagerInfo(sameHostBlockId)
+            bmInfo.blockManagerId -> bmInfo.localDirs
+          }
+      } else {
+        None
+      }
+      bmIdToLocalDirs.map { case (bmId, localDirs) =>
+        Some(BlockLocationsAndStatus(locations.filter(_ != bmId), status.get, Some(localDirs)))
+      }.getOrElse(
+        Some(BlockLocationsAndStatus(locations, status.get, None))
```

Comment:
minor: instead of `getOrElse`, you can use `orElse`, to avoid some of the extra `Some()` wrapping:

```scala
scala> def foo(x: Option[Int]): Option[String] = {
     |   x.map { i => i.toString}.orElse(Some("nope"))
     | }
foo: (x: Option[Int])Option[String]

scala> 

scala> foo(None)
res1: Option[String] = Some(nope)

scala> foo(Some(1))
res2: Option[String] = Some(1)
```